### PR TITLE
Support mixed basic and advanced indexing in ops.index_update

### LINKED ISSUE #988: Implement indexed updates that mix basic and advanced indexing
I'm trying to update certain rows of a matrix with corresponding gradients. I tried "index_add" and it raised NotImplementedError for _scattter_update error for now. This is crucial for my use case where weight matrix selection is data-dependent. 

  ```new_weights = index_add(weights, ops.index[neuron_indices + context_values, :], - learning_rate * gradient)```

### PR DESCRIPTION
**Pull request by hawkinsp**
FYI: I have a follow-on PR that merges the gather and scatter implementations (by replacing most of the `_rewriting_take` implementation with the scatter implementation, which is now shared with `_rewriting_take` and adapted to call gather instead of scatter):
https://github.com/hawkinsp/jax/compare/scatter...hawkinsp:master

Sending this part out first separately because it a less complicated change to make and I have more confidence it won't break things.

Fixes #988

### CODE REVIEW

**Review by mattjj — APPROVED** [review]
LGTM!

I'm curious: how long does lax_numpy_indexing_test.py take to run before and after this change? Any difference?

**Comment by hawkinsp** [comment]
There seems to be no impact on the test time (for `JAX_NUM_GENERATED_CASES=10`) for either change. I'm a bit suspicious of that — it might simply say the tests have some sort of large fixed overhead. That said, the index-update change seems safe enough either way.
